akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala

```diff
@@ -875,12 +873,12 @@ final class Partition[T](val outputPorts: Int, val partitioner: T => Int, val ea
                   downstreamRunning -= 1
                   if (downstreamRunning == 0)
                     cancelStage(cause)
-                  else if (outPendingElem != null) {
-                    if (idx == outPendingIdx) {
-                      outPendingElem = null
-                      if (!hasBeenPulled(in))
-                        pull(in)
-                    }
+                  else if (outPendingElem != null && idx == outPendingIdx) {
+                    outPendingElem = null
+                    if (isClosed(in))
+                      cancelStage(cause)
```

Comment:
Interesting that when downstream cancels with an error, it's now not very predictable whether that cause will propagate or be dropped - but that's outside the scope of this PR